helix-core/src/main/java/org/apache/helix/util/WeightAwareRebalanceUtil.java

```diff
@@ -0,0 +1,141 @@
+package org.apache.helix.util;
+
+import org.apache.helix.HelixException;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.api.config.RebalanceConfig;
+import org.apache.helix.api.rebalancer.constraint.AbstractRebalanceHardConstraint;
+import org.apache.helix.api.rebalancer.constraint.AbstractRebalanceSoftConstraint;
+import org.apache.helix.controller.common.PartitionStateMap;
+import org.apache.helix.controller.common.ResourcesStateMap;
+import org.apache.helix.controller.rebalancer.strategy.ConstraintRebalanceStrategy;
+import org.apache.helix.controller.stages.ClusterDataCache;
+import org.apache.helix.model.*;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A rebalance tool that generate an resource partition assignment based on the input.
+ * Note the assignment won't be automatically applied to the cluster. Applications are supposed to
+ * apply the change.
+ */
+public class WeightAwareRebalanceUtil {
+  private final ClusterConfig _clusterConfig;
+  private final Map<String, InstanceConfig> _instanceConfigMap = new HashMap<>();
+  // For the possible customized state models.
+  private final Map<String, StateModelDefinition> _stateModelDefs = new HashMap<>();
+  private final ClusterDataCache _dataCache;
+
+  public enum RebalanceOption {
+    INIT,
+    REASSIGN
+  }
+
+  public WeightAwareRebalanceUtil(ClusterConfig clusterConfig,
+      List<InstanceConfig> instanceConfigs) {
+    for (InstanceConfig instanceConfig : instanceConfigs) {
+      _instanceConfigMap.put(instanceConfig.getInstanceName(), instanceConfig);
+    }
+    _clusterConfig = clusterConfig;
+
+    _dataCache = new ClusterDataCache();
+    _dataCache.setInstanceConfigMap(_instanceConfigMap);
+    _dataCache.setClusterConfig(_clusterConfig);
+    List<LiveInstance> liveInstanceList = new ArrayList<>();
+    for (String instance : _instanceConfigMap.keySet()) {
+      LiveInstance liveInstance = new LiveInstance(instance);
+      liveInstanceList.add(liveInstance);
+    }
+    _dataCache.setLiveInstances(liveInstanceList);
+  }
+
+  /**
+   * The method to generate partition assignment mappings.
+   *
+   * @param resourceConfigs    Config of all the resources that need to be rebalanced.
+   *                           The tool throws Exception if any resource has no IS or broken/uninitialized IS.
+   *                           The tool throws Exception if any resource is in full-auto mode.
+   * @param existingAssignment The existing partition assignment of the input resources.
+   * @param option             INIT or REASSIGN
+   *                           INIT: Keep existing assignment. Only generate new partition assignment.
```

Comment:
instead of option, can we call it 'mode'? Also, it is not clear where we specify new partitions to be assigned. "INIT" also seems to be misleading. Why not "KEEP"? If "INIT" is specified, then is it true that the existing assignments will not be moved? if "REASSIGN" is specified, is it true that the algorithm will try not to move existing assignments,  but may do so?

Reply:
"Mode" is confusing while we already have "RebalanceMode".

About the option name, how about "incremental" rebalance & "full" rebalance? I think those 2 names are much better.

I will add the comment about the resourceConfig, this is where users should specify the partitions.
Also, as a standalone tool, most rebalance options won't work even configured in the cluster config / instance config. I will add comments to point it out.

Reply:
And about the last point. You are correct. Since the rebalance algorithm is optimized to minimize movement, even with full rebalance, the new mapping is supposed to keep as much same assignment as the old one. Just not guaranteed.

Reply:
If you are having difficulty naming the variable, use two different APIs.

buildIncrementalRebalanceAssignment()
buildFullRebalanceAssignment()

Or some such.

I still don't understand the difference between the two. 

I suppose incremental means to calculate idealstate assignment for new partitions, given the old partition assignment. Right? In that case the list of new partitions should be taken as an argument, but it is not. Old partitions are not touched. Am I right?

I suppose full rebalance means to re-assign existing partitions around (but no new partitions added). Am I right? In  this case, we rebalance the partitions, but minimize movement. Am I right?

Reply:
You are right, Subbu.
The existingAssignment is old partition assignment.
For incremental, those old assignments won't be changed unless they are not valid (compared with State Model requirement).
For full rebalance, the old assignments are only used to minimize movement if possible. But all existing assignment will be recalculated.

And totally agree with your suggestion that make 2 separate APIs. This also helps clean the javadoc for parameters.
I will do it. Thanks for the comment.